Replace QuaZip (and other archiving operations) with libarchive

Trial97 opened the pull request:
need to test the issues that were fixed by:
- https://github.com/PrismLauncher/PrismLauncher/pull/2225
- https://github.com/PrismLauncher/PrismLauncher/pull/2353

Other stuff to test:
- [x] import instance 
  - [x] Curseforge
  - [x] Modrinth
  - [x] MultiMc
  - [x] AtLaucnher
  - [x] Technic
  - [x] FTB
- [x] java download
- [x] native extracting(I do not know what and when this)
- [x] resource parsing (some have tests)
  - [x] data pack
  - [x] mod
  - [x] shader
  - [x] texture
  - [x] world
- [x] export instance
  - [x] Curseforge
  - [x] Modrinth
  - [x] Prismlauncher 
- [x] check the locales stuff (i set the locales each time I try to extract/read a archive)

Stuff to improve:
- [x] ~~mod parsing seems slower(and it may be as it reopens the jars a few times)~~ It should be fixed. If not too bad as the library doesn't allow for faster reading from what I know.

@getchoo I need a bit of help  with the action building stuff, please advise 


# Today I learned: libarchive pollutes the code source; avoid using the header(< archive.h>) or any header from that library in our own headers, keep them strictly to the cpp files. 
On windows I got `C2143: syntax error: missing '}' before ','
D:\a\PrismLauncher\PrismLauncher\launcher\modplatform/ModIndex.h(48): error C2059: syntax error: ','` Because the `UNKNOWN` is defined somewhere in the libarchive code.

blocked on #3977

Trial97 committed:
move ExportToZipTask

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

getchoo commented on CMakeLists.txt:
```diff
@@ -362,6 +362,15 @@ if(NOT Launcher_FORCE_BUNDLED_LIBS)
+        if(PkgConfig_FOUND)
+            pkg_check_modules(LibArchive IMPORTED_TARGET LibArchive>=3.7.8)
+        endif()
+    endif()
```
```suggestion
    find_package(PkgConfig REQUIRED)
    pkg_check_modules(libarchive REQUIRED IMPORTED_TARGET libarchive)
```
I don't know of any distribution of libarchive that has CMake support...

getchoo commented on launcher/CMakeLists.txt:
```diff
@@ -1320,6 +1324,12 @@ else()
+    target_link_libraries(Launcher_logic PkgConfig::LibArchive)
+else()
+    target_link_libraries(Launcher_logic LibArchive::LibArchive)
+endif()
```
```suggestion
```
Include `PkgConfig::libarchive` above

Trial97 committed:
move some functions from MMCZip to use libarchive

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

getchoo commented on CMakeLists.txt:
```diff
@@ -346,14 +346,13 @@ if(Launcher_QT_VERSION_MAJOR EQUAL 6)
+    pkg_check_modules(libarchive IMPORTED_TARGET libarchive)
     # Find toml++
     find_package(tomlplusplus 3.2.0 QUIET)
     # Fallback to pkg-config (if available) if CMake files aren't found
```
```suggestion
    # Fallback to pkg-config if CMake files aren't found
```

Trial97 committed:
move ExtractZipTask

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
move more zip parsings

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
finish moving all code to libarchive

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
remove quazip

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
add missing copyrights

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
add libarchive as actions dependency

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

getchoo commented on CMakeLists.txt:
```diff
@@ -346,14 +338,13 @@ if(Launcher_QT_VERSION_MAJOR EQUAL 6)
 
 if(NOT Launcher_FORCE_BUNDLED_LIBS)
+    find_package(PkgConfig REQUIRED)
+    pkg_check_modules(libarchive IMPORTED_TARGET libarchive)
```
This should also be taken out of the current conditional since no bundled version of this exists
```suggestion
    find_package(PkgConfig REQUIRED)
    pkg_check_modules(libarchive REQUIRED IMPORTED_TARGET libarchive)
```

Trial97 committed:
move libarchive outside bundle libs

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
Improve mod parsing

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
fix nix build

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
build(cmake): better detect libarchive

Co-authored-by: Octol1ttle <l1ttleofficial@outlook.com>
Co-authored-by: Seth Flynn <getchoo@tuta.io>
Signed-off-by: Alexandru Ionut Tripon <alexandru.tripon97@gmail.com>
Signed-off-by: Octol1ttle <l1ttleofficial@outlook.com>
Signed-off-by: Seth Flynn <getchoo@tuta.io>
Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

getchoo commented on CMakeLists.txt:
```diff
@@ -345,15 +337,23 @@ if(Launcher_QT_VERSION_MAJOR EQUAL 6)
+    pkg_check_modules(libarchive REQUIRED IMPORTED_TARGET libarchive)
+    set(LIBARCHIVE_LIBRARY PkgConfig::libarchive)
+else()
+    find_library(LibArchive REQUIRED)
+    set(LIBARCHIVE_LIBRARY LibArchive::LibArchive)
```
```suggestion
    find_path(LIBARCHIVE_INCLUDE_DIR archive.h REQUIRED)
    find_library(LIBARCHIVE_LIBRARY_RELEASE archive REQUIRED)
    find_library(LIBARCHIE_LIBRARY_DEBUG archived)
    set(LIBARCHIVE_LIBRARIES optimized ${LIBARCHIVE_LIBRARY_RELEASE} debug ${LIBARCHIVE_LIBRARY_DEBUG})
```

getchoo commented on CMakeLists.txt:
```diff
@@ -345,15 +337,23 @@ if(Launcher_QT_VERSION_MAJOR EQUAL 6)
 endif()
 
+find_package(PkgConfig REQUIRED)
+# Find libqrencode
+## NOTE(@getchoo): Never use pkg-config with MSVC since the vcpkg port makes our install bundle fail to find the dll
+if(PkgConfig_FOUND AND NOT MSVC)
+    pkg_check_modules(libarchive REQUIRED IMPORTED_TARGET libarchive)
+    set(LIBARCHIVE_LIBRARY PkgConfig::libarchive)
```
```suggestion
# Find libarchive
## NOTE(@getchoo): Never use pkg-config with MSVC since the vcpkg port makes our install bundle fail to find the dll
if(PkgConfig_FOUND AND NOT MSVC)
    pkg_check_modules(libarchive REQUIRED IMPORTED_TARGET libarchive)
    set(LIBARCHIVE_LIBRARIES PkgConfig::libarchive)
```

getchoo commented on CMakeLists.txt:
```diff
@@ -345,15 +337,23 @@ if(Launcher_QT_VERSION_MAJOR EQUAL 6)
     set(QT_LIBEXECS_DIR ${QT${QT_VERSION_MAJOR}_INSTALL_PREFIX}/${QT${QT_VERSION_MAJOR}_INSTALL_LIBEXECS})
 endif()
 
+find_package(PkgConfig REQUIRED)
```
```suggestion
find_package(PkgConfig)
```
This technically isn't required since we have a fallback

getchoo commented on CMakeLists.txt:
```diff
@@ -345,15 +337,23 @@ if(Launcher_QT_VERSION_MAJOR EQUAL 6)
-        if(PkgConfig_FOUND)
-            pkg_check_modules(tomlplusplus IMPORTED_TARGET tomlplusplus>=3.2.0)
-        endif()
+        pkg_check_modules(tomlplusplus IMPORTED_TARGET tomlplusplus>=3.2.0)
```
```suggestion
        if(PkgConfig_FOUND)
            pkg_check_modules(tomlplusplus IMPORTED_TARGET tomlplusplus>=3.2.0)
        endif()
```
Ditto

getchoo commented on CMakeLists.txt:
```diff
@@ -345,15 +337,23 @@ if(Launcher_QT_VERSION_MAJOR EQUAL 6)
     # Find toml++
     find_package(tomlplusplus 3.2.0 QUIET)
-    # Fallback to pkg-config (if available) if CMake files aren't found
+    # Fallback to pkg-config if CMake files aren't found
```
```suggestion
    # Fallback to pkg-config (if available) if CMake files aren't found
```

getchoo commented on launcher/CMakeLists.txt:
```diff
@@ -1322,6 +1332,7 @@ target_link_libraries(Launcher_logic
     BuildConfig
     Qt${QT_VERSION_MAJOR}::Widgets
     qrcodegenerator
+    ${LIBARCHIVE_LIBRARY}
```
```suggestion
    ${LIBARCHIVE_LIBRARIES}
```

getchoo commented on launcher/CMakeLists.txt:
```diff
@@ -1424,6 +1433,7 @@ if(Launcher_BUILD_UPDATER)
         Qt${QT_VERSION_MAJOR}::Network
         ${Launcher_QT_LIBS}
         cmark::cmark
+        ${LIBARCHIVE_LIBRARY}
```
```suggestion
        ${LIBARCHIVE_LIBRARIES}
```

getchoo commented on CMakeLists.txt:
```diff
@@ -345,12 +337,24 @@ if(Launcher_QT_VERSION_MAJOR EQUAL 6)
+else()
+    find_path(LIBARCHIVE_INCLUDE_DIR archive.h REQUIRED)
+    find_library(LIBARCHIVE_LIBRARY_RELEASE archive REQUIRED)
+    find_library(LIBARCHIE_LIBRARY_DEBUG archived)
+    set(LIBARCHIVE_LIBRARIES optimized ${LIBARCHIVE_LIBRARY_RELEASE} debug ${LIBARCHIVE_LIBRARY_DEBUG})
```
```suggestion
    set(LIBARCHIVE_LIBRARIES optimized ${LIBARCHIVE_LIBRARY_RELEASE})
```
Seems libarchive doesn't generate these

getchoo commented on .github/actions/setup-dependencies/macos/action.yml:
```diff
@@ -14,7 +14,7 @@ runs:
       run: |
         brew update
-        brew install ninja extra-cmake-modules temurin@17
+        brew install ninja extra-cmake-modules temurin@17 libarchive
```
```suggestion
        brew install ninja extra-cmake-modules temurin@17
```
vcpkg is used instead now

Trial97 committed:
fix merge conflicts

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

getchoo committed:
build(vcpkg): remove bzip2 dependency

This was previously required by Quazip

Signed-off-by: Seth Flynn <getchoo@tuta.io>
Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

TayouVR commented on launcher/archive/ArchiveReader.cpp:
```diff
@@ -0,0 +1,244 @@
+ *      distributed under the License is distributed on an "AS IS" BASIS,
+ *      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *      See the License for the specific language governing permissions and
+ *      limitations under the License.
```
^

TayouVR commented on launcher/archive/ArchiveReader.h:
```diff
@@ -0,0 +1,88 @@
+ *      distributed under the License is distributed on an "AS IS" BASIS,
+ *      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *      See the License for the specific language governing permissions and
+ *      limitations under the License.
```
^

TayouVR commented on launcher/archive/ArchiveWriter.cpp:
```diff
@@ -0,0 +1,230 @@
+ *      distributed under the License is distributed on an "AS IS" BASIS,
+ *      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *      See the License for the specific language governing permissions and
+ *      limitations under the License.
```
^

TayouVR commented on launcher/archive/ExportToZipTask.cpp:
```diff
@@ -0,0 +1,117 @@
+ *      distributed under the License is distributed on an "AS IS" BASIS,
+ *      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *      See the License for the specific language governing permissions and
+ *      limitations under the License.
```
^

TayouVR commented on launcher/archive/ExportToZipTask.h:
```diff
@@ -0,0 +1,89 @@
+ *      distributed under the License is distributed on an "AS IS" BASIS,
+ *      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *      See the License for the specific language governing permissions and
+ *      limitations under the License.
```
^

TayouVR commented on launcher/archive/ExtractZipTask.cpp:
```diff
@@ -0,0 +1,152 @@
+ *      distributed under the License is distributed on an "AS IS" BASIS,
+ *      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *      See the License for the specific language governing permissions and
+ *      limitations under the License.
```
^

TayouVR commented on launcher/archive/ExtractZipTask.h:
```diff
@@ -0,0 +1,70 @@
+ *      distributed under the License is distributed on an "AS IS" BASIS,
+ *      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *      See the License for the specific language governing permissions and
+ *      limitations under the License.
```
^

TayouVR reviewed (commented):
some of the files are entirely new, but have license headers mentioning MultiMC and its license still. For the code thats new thats not necessary I'd say.

Trial97 committed:
fix license headers

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Octol1ttle commented on libraries/quazip:
Also delete the license notice in `libraries/README.md`

Trial97 committed:
update license readme

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
fix: CF import and Modrinth export

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 commented:
Tested all the mentioned checkboxes above and also tested if I can import the instance once exported in CF launcher
All tests were done on linux.

Octol1ttle commented:
Testing on Windows MSVC, found a regression (#1497). Used all 3 export options to export from Prism, then imported them back
Top: imported, bottom: exported (expected result)
<img width="547" height="358" alt="изображение" src="https://github.com/user-attachments/assets/f6a9711d-d08d-4a27-a907-fa4880b4ce36" />

Java download and native extracting appears to work fine, launcher managed to download everything and the game launched normally, narrator works.
Downloading and parsing of custom resources works good too.

Trial97 committed:
fix: utf8 pathname encoding

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 commented:
@Octol1ttle can you please retest with the latest change?

Trial97 commented:
@Octol1ttle can you see if you can reproduce the crash now that we also set the locale
Also added to the tests is the `Use system locales` settings from the Language settings

Octol1ttle commented:
> @Octol1ttle can you see if you can reproduce the crash now that we also set the locale Also added to the tests is the `Use system locales` settings from the Language settings

there's a different crash now
<img width="412" height="315" alt="изображение" src="https://github.com/user-attachments/assets/368c66fa-c0bc-4b9e-9a1c-2b27c819ddc9" />
debugger view:
<img width="1242" height="403" alt="изображение" src="https://github.com/user-attachments/assets/a6cef00f-6c29-4b97-980e-96fb37ba334f" />
last relevant log lines:
```
    16.661 D: [launcher.task] Task "MMCZip::ExportToZipTask(0x1ca67cc7b50 ID: c84f9e70-673f-4646-8721-1da2c75c5bbc)" starting for the first time (Task::start:80)
    16.875 C: Failed to write header for:  "minecraft/saves/TEST قلب 你好 Привет नमस्ते ελληνι/advancements/ac5427e5-9f37-44be-b239-de5c44d0f25b.json" (MMCZip::ArchiveWriter::addFile:141)
    16.999 W: Failed to remove files: "The process cannot access the file because it is being used by another process." (FS::deletePath:679)
    17.001 C: [launcher.task] Task "MMCZip::ExportToZipTask(0x1ca67cc7b50 ID: c84f9e70-673f-4646-8721-1da2c75c5bbc)" failed:  "Не удалось прочитать и сжать minecraft/saves/TEST قلب 你好 Привет नमस्ते ελληνι/advancements/ac5427e5-9f37-44be-b239-de5c44d0f25b.json" (Task::emitFailed:119)
    32.980 D: [launcher.task.net.metacache.http] Saving metacache with 14 entries (HttpMetaCache::SaveNow:295)
Exception: Exception 0x80000003 encountered at address 0x7ff82ae08207
```
`Failed to write header` error code: `-20`

Octol1ttle commented:
Testing on Windows MSVC again. Exporting and importing instances with special symbols works as intended. Resource downloading (Java, libraries, natives, custom user data such as resourcepacks), unpacking and parsing also works

Trial97 committed:
ensure secure extraction

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
nothing but zip writes

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
update the vcpkg baseline

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

getchoo committed:
build(vcpkg): minimize libarchive build

This avoids pulling in libxml2 for xar support, which we don't use and
was giving us some trouble in CI

Signed-off-by: Seth Flynn <getchoo@tuta.io>
Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

DioEgizio commented:
<img width="225" height="106" alt="image" src="https://github.com/user-attachments/assets/a0e12f72-d599-44fe-b852-67374c708bbd" />
(on linux) importing and exporting with special characters works great, however exporting completely kills timestamps. This is a regression compared to before

getchoo commented:
We're also likely losing other attributes of files like permissions, flags, or any other kind of ACLs since we aren't using their equivalent `archive_entry` functions

Trial97 committed:
use file stat to get time information

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

getchoo commented:
> We're also likely losing other attributes of files like permissions, flags, or any other kind of ACLs since we aren't using their equivalent `archive_entry` functions

Still an issue, and likely to affect instances that have anything executable in them for example

Trial97 commented:
Pretty sure I set the permissions correctly now

getchoo commented on launcher/archive/ArchiveWriter.cpp:
```diff
@@ -0,0 +1,211 @@
+    if (stat(cpath, &st) == 0) {
+        archive_entry_copy_stat(entry, &st);
+    }
+    archive_entry_set_perm(entry, fileInfo.permissions() & 0777);
```
```suggestion
    archive_entry_set_perm(entry, fileInfo.permissions());
```
Pretty sure masking this value would lose permission attributes for NTFS/Windows stuff

Trial97 replied:
nope this one only loses SUID/SGID/sticky bits not the rwx permissions

Trial97 replied:
but I see no reason not to follow your suggestion

Trial97 committed:
Update launcher/archive/ArchiveWriter.cpp

Co-authored-by: Seth Flynn <getchoo@tuta.io>
Signed-off-by: Alexandru Ionut Tripon <alexandru.tripon97@gmail.com>

Trial97 committed:
fix merge

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Ryex reviewed (approved):
We probably want to send a macos test build out but LGTM and testing went fine

DioEgizio commented:
just tested, everything works fine on macOS

getchoo commented on launcher/archive/ArchiveWriter.cpp:
```diff
@@ -0,0 +1,211 @@
+        return false;
+    }
+
+    archive_write_set_format_zip(m_archive);
```
```suggestion
    archive_write_set_format_by_name(m_archive, m_format);
```
Small change to make writing other formats actually possible

getchoo commented on launcher/archive/ArchiveWriter.h:
```diff
@@ -0,0 +1,45 @@
+
+   private:
+    struct archive* m_archive = nullptr;
+    QString m_filename;
```
```suggestion
    QString m_filename;
    QString m_format = "zip";
```
But keeping with zip as a default

getchoo commented on launcher/archive/ArchiveWriter.cpp:
```diff
@@ -0,0 +1,211 @@
+{
+    QFileInfo fileInfo(fileName);
+    if (!fileInfo.exists()) {
+        qCritical() << "File does not exists:" << fileInfo.filePath();
```
```suggestion
        qCritical() << "File does not exist:" << fileInfo.filePath();
```

getchoo commented on launcher/archive/ArchiveWriter.cpp:
```diff
@@ -0,0 +1,211 @@
+    struct stat st;
+    if (stat(cpath, &st) == 0) {
+        archive_entry_copy_stat(entry, &st);
+    }
```
```suggestion
    if (stat(cpath, &st) != 0) {
        qCritical() << "Failed to stat file:" << fileInfo.filePath();
    }
    archive_entry_copy_stat(entry, &st);
```
Being reliant on `stat` should be able to save us a few steps later - and after all, I don't know why this would actually fail in any reasonable case...

getchoo commented on launcher/archive/ArchiveWriter.cpp:
```diff
@@ -0,0 +1,211 @@
+        archive_entry_copy_stat(entry, &st);
+    }
+    archive_entry_set_perm(entry, fileInfo.permissions());
+
+    if (fileInfo.isSymLink()) {
+        auto target = fileInfo.symLinkTarget().toUtf8();
+        archive_entry_set_filetype(entry, AE_IFLNK);
+        archive_entry_set_symlink(entry, target.constData());
+        archive_entry_set_size(entry, 0);
```
```suggestion
        archive_entry_set_symlink(entry, target.constData());
```
These *should* be handled by stat - or at least, it doesn't make a difference on Windows and it makes sense by the docs

getchoo commented on launcher/archive/ArchiveWriter.cpp:
```diff
@@ -0,0 +1,211 @@
+        return true;
+    }
+
+    if (!fileInfo.isFile()) {
```
```suggestion
    if (!fileInfo.isFile() && !fileInfo.isDir()) {
```
QuaZip didn't do this, but I think it could be nice to handle directories here. Would make it easy for people to have users drop stuff into readily made folders with exported instances, for example

getchoo commented on launcher/archive/ArchiveWriter.cpp:
```diff
@@ -0,0 +1,211 @@
+    }
+
+    archive_entry_set_filetype(entry, AE_IFREG);
+    archive_entry_set_size(entry, file.size());
```
We should be able to remove this, as it should also be copied over from `stat()`

getchoo commented on launcher/archive/ArchiveWriter.cpp:
```diff
@@ -0,0 +1,211 @@
+    if (!file.open(QIODevice::ReadOnly)) {
+        qCritical() << "Failed to open file: " << fileInfo.filePath();
+        return false;
+    }
```
We could then move this down a bit

```suggestion
```

getchoo commented on launcher/archive/ArchiveWriter.cpp:
```diff
@@ -0,0 +1,211 @@
+            qCritical() << "Write error in archive for: " << fileDest;
+            return false;
+        }
+    }
```
And write the header for all files, directories, and symlinks above, while only writing the data for regular files here

```suggestion
    if (fileInfo.isFile() && !fileInfo.isSymLink()) {
        QFile file(fileInfo.absoluteFilePath());
        if (!file.open(QIODevice::ReadOnly)) {
            qCritical() << "Failed to open file:" << fileInfo.filePath();
            return false;
        }

        constexpr qint64 chunkSize = 8192;
        QByteArray buffer;
        buffer.resize(chunkSize);
    
        while (!file.atEnd()) {
            auto bytesRead = file.read(buffer.data(), chunkSize);
            if (bytesRead < 0) {
                qCritical() << "Read error in file: " << fileInfo.filePath();
                return false;
            }
    
            if (archive_write_data(m_archive, buffer.constData(), bytesRead) < 0) {
                qCritical() << "Write error in archive for: " << fileDest;
                return false;
            }
        }
    }
```

Trial97 replied:
here I do not need the check for a symlink as I exit early above if this happens

getchoo commented on launcher/archive/ArchiveReader.cpp:
```diff
@@ -0,0 +1,227 @@
+    return nullptr;
+}
+
+static int copy_data(struct archive* ar, struct archive* aw, bool notBlock = false)
+{
+    int r;
+    const void* buff;
+    size_t size;
+    la_int64_t offset;
+
+    for (;;) {
+        r = archive_read_data_block(ar, &buff, &size, &offset);
+        if (r == ARCHIVE_EOF)
+            return (ARCHIVE_OK);
+        if (r < ARCHIVE_OK) {
+            qCritical() << "Failed reading data block:" << archive_error_string(ar);
+            return (r);
+        }
+        if (notBlock) {
+            r = archive_write_data(aw, buff, size);
+        } else {
+            r = archive_write_data_block(aw, buff, size, offset);
+        }
+        if (r < ARCHIVE_OK) {
+            qCritical() << "Failed writing data block:" << archive_error_string(aw);
+            return (r);
+        }
+    }
+}
```
This looks like a modification of [code from libarchive's examples](https://github.com/libarchive/libarchive/blob/8ef2084104887af3401d91b36eab2496767c81a8/examples/untar.c#L211-L223). Might be worth adding this to the SDPX header under `LicenseRef-PublicDomain`

DioEgizio replied:
it's literally public domain it doesn't matter

getchoo replied:
That's still a license to consider, though. Distributions like Fedora *do* include this information

[We also do this in other cases of public domain work](https://github.com/PrismLauncher/PrismLauncher/blob/5d1e457b3904c3b39e2227678f8f4034668927b5/libraries/murmur2/src/MurmurHash2.h#L2-L3) - though that's a more extreme case obviously

DioEgizio replied:
> That's still a license to consider, though. Distributions like Fedora *do* include this information
> 
> [We also do this in other cases of public domain work](https://github.com/PrismLauncher/PrismLauncher/blob/5d1e457b3904c3b39e2227678f8f4034668927b5/libraries/murmur2/src/MurmurHash2.h#L2-L3) - though that's a more extreme case obviously

i guess so but public domain doesn't require us to do literally anything so it's weird to show this as a license. Still if some distros some fedora do it it's not as dumb as I thought.

Trial97 replied:
Please do check if the header matches with wat you wanted

getchoo commented on launcher/archive/ArchiveReader.cpp:
```diff
@@ -0,0 +1,227 @@
+    }
+    auto r = archive_write_finish_entry(out);
+    if (r < ARCHIVE_OK)
+        qCritical() << "Failed dinish entry:" << archive_error_string(out);
```
```suggestion
        qCritical() << "Failed to finish writing entry:" << archive_error_string(out);
```

getchoo commented on launcher/archive/ArchiveReader.cpp:
```diff
@@ -0,0 +1,227 @@
+    archive_read_support_format_all(a);
+    archive_read_support_filter_all(a);
+    auto fileName = m_archivePath.toUtf8();
+    if (archive_read_open_filename(a, fileName.constData(), 10240) != ARCHIVE_OK) {
```
The block size would be nice as a const

getchoo reviewed (commented):
Few nitpicks and code cleanups

[patch file to save you time](https://github.com/user-attachments/files/23775658/0001-refactor-launcher-archive-extend-archive-write-suppo.patch)

getchoo reviewed (approved):
~~this was sent in the wrong tab~~

getchoo reviewed (dismissed):
oopsie

Trial97 committed:
apply suggestions

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
add back kdePackages.qt5compat

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

getchoo committed:
refactor(launcher/archive): ensure correct filetype for archived files

We can rely on stat for most things but this

Signed-off-by: Seth Flynn <getchoo@tuta.io>

getchoo reviewed (approved):
I'm very happy with this. Thanks for all your work here, Trial :)

Scrumplex reviewed (approved):
cool

getchoo commented:
<img width="1162" height="783" alt="image" src="https://github.com/user-attachments/assets/aa6ba2d8-8b63-405b-82aa-9c421883eca7" />
